[FEATURE] Use TIFFTAG_DOCUMENTNAME as layer name if available

oparkins opened the pull request:
## Description

Originally the GDAL data provider ignores any metadata associated with TIFF files. Many documents contain metadata that provides the name and description of the file. This is more important to provide to the user when dealing with sublayers where users are unable to provide a name externally by renaming the file.

This commit allows for sublayers to be defined by this metadata and provided to the user for sublayer selection. If it doesn't exist, it performs the default behavior.

I only modified this for TIFF files with sublayers because in my limited experience, users expect the single layer files to be imported with the same name as found on the operating system. This model breaks down with sublayers since the user cannot provide a sane name for each sublayer.

Related to qgis/QGIS#57077

In the issue mentioned above, I originally stated I didn't think this changed needed to be implemented in the `SUBDATASET` querying code. I changed my approach once I started to dig into this a bit more and found that the `name` was not used as the `uri` for the sublayer location. This enable the change to be made only in one spot and affect the display of the layer elsewhere.


---

Here is what the import originally looks like for a multi-page tiff file:

![before_sublayer_dialog](https://github.com/qgis/QGIS/assets/16523126/2af09773-18fc-4bc2-b071-91d54c3d5638)

And the default layers list once imported:

![before_layers_list](https://github.com/qgis/QGIS/assets/16523126/401e0615-cd40-4e57-ab8b-90507b44bfd9)


And now:

![after_sublayer_dialog](https://github.com/qgis/QGIS/assets/16523126/62532645-493e-4699-8f09-fd9d78a1d68e)

And the default layers list once imported:

![after_layers_list](https://github.com/qgis/QGIS/assets/16523126/b622723c-5c48-4046-9512-78355a86bee6)

---

Todo:
- [x] Add unit test with sublayers
- [x] Ensure scripts are ran before commits as stated in PR help text
- [x] Check coding standards document as mentioned in the PR help text


For the tests, I just added another example to the repo that mirrored the `gtiff_tags` test. The test that I modified still passes. 

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

oparkins committed:
[FEATURE] Use TIFFTAG_DOCUMENTNAME as layer name if available

Originally the GDAL data provider ignores any metadata
associated with tiff documents. Many documents contain
metadata that provides the name and description of the
document. This is more important to provide to the user
 when dealing with sublayers where users are unable to
provide a name externally by renaming the file.

This commit allows for sublayers to be defined by
this metadata and provided to the user for sublayer
selection.

Related to qgis/QGIS#57077

oparkins committed:
[STYLE] Updated local variables to reflect code standards

- Removed the prefix 'my' from local variables
- Ran the prepare_commit.sh script to fix style

oparkins committed:
[TEST] Add tests for TIFF Tags

- Added a test raster that contains multiple layers with the
  appropriate tag for naming the image and adding the description
- Added a test to ensure the names are used when available

Related to qgis/QGIS#57146

oparkins commented:
This should be good for someone to look at now. Really appreciate the structure of QGIS and the code quality since it was easy to implement this feature. 

Let me know if there is anything I should fix/change/etc. Looking forward to having this available for our multi-page TIFF files we generate.

rouault commented on src/core/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
         {
+
+          // Check if the layer has TIFFTAG_DOCUMENTNAME associated with it. If so, use that name.
+          GDALDatasetH datasetHandle = GDALOpen( name, GA_ReadOnly );
```
below code addition should only be execute under ``if (datasetHandle)``

oparkins replied:
Good point! Thanks!

rouault commented on src/core/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
+          // Check if the layer has TIFFTAG_DOCUMENTNAME associated with it. If so, use that name.
+          GDALDatasetH datasetHandle = GDALOpen( name, GA_ReadOnly );
+          QString tagTIFFDocumentName = GDALGetMetadataItem( datasetHandle, "TIFFTAG_DOCUMENTNAME", nullptr );
+          if ( tagTIFFDocumentName != nullptr )
```
tagTIFFDocumentName is a QString, not a pointer, so can' be nullptr:

```suggestion
          if ( !tagTIFFDocumentName.isEmpty() )
```

rouault commented on src/core/providers/gdal/qgsgdalprovider.cpp:
```diff
@@ -1835,12 +1835,30 @@ QList<QgsProviderSublayerDetails> QgsGdalProvider::sublayerDetails( GDALDatasetH
+          }
+
+          QString tagTIFFImageDescription = GDALGetMetadataItem( datasetHandle, "TIFFTAG_IMAGEDESCRIPTION", nullptr );
+          if ( tagTIFFImageDescription != nullptr )
```
```suggestion
          if ( !tagTIFFImageDescription.isEmpty() )
```

oparkins committed:
[fix] Updated tests for GDAL Handle and QString

@rouault reviewed the code and mentioned that QString needs to be
checked if empty in a different way than nullptr. Also mentioned
that the GDAL Handle needed to be checked for validity. Thanks!

Also fixed the test because there was a left over import upon
a clean rebuild. This has been removed

oparkins committed:
[test] Fixed the test in this commit

Mentioned in previous commit about test needing to be fixed. Forgot
to add it to the previous commit.

oparkins commented:
Thanks for the review @rouault. I've made those changes.

oparkins committed:
[test] Fixed and provided better test

- Fixed test URI path
- Updated test raster to contain unique values for
  each metadata item. This is reflected in the test
  now. This means the test is also checking that it
  is reading the right metadata from the sublayer.

Related to qgis#57146

oparkins commented:
Just pushed a fix for those tests. Proof of them working on my machine now:

```bash
ctest -R test_core_gdalprovider
Test project /home/oparkins/projects/QGIS/build-master
    Start  1: logGitStatus
1/3 Test  #1: logGitStatus .....................   Passed    0.05 sec
    Start 70: test_core_gdalprovider
2/3 Test #70: test_core_gdalprovider ...........   Passed    1.01 sec
    Start  2: checkGitStatus
3/3 Test  #2: checkGitStatus ...................   Passed    0.04 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   1.14 sec
```

Not sure how I messed that up, but I did. This should have fixed that.

oparkins committed:
Merge branch 'qgis:master' into gdal-tiff-use-tags-for-name

oparkins commented:
I just synced my fork to make future merges easier. I'm also hoping that will enable the workflows to pass. The last round of failures seem to be caused by another test case that is not relevant to my changes.

rouault commented:
so we have those failures:
```
 	136 - test_core_networkaccessmanager (Timeout)
	356 - test_app_maptooladdpart (Failed)
```
The first one is likely a random flake (I'm restarting the build to test that), and the second is definitely not due to this pull request, as occuring in master builds too (https://github.com/qgis/QGIS/actions/runs/8913989856/job/24481614072)

oparkins commented:
Awesome. Thanks for checking into that. Glad I wasn't going too crazy.

Besides those two failed tests, this PR should contain all that is needed to close the associated issue https://github.com/qgis/QGIS/issues/57077. If there is anything that you see needs correction related to this PR, more than happy to keep working on it. Just let me know how I can be helpful.
